### PR TITLE
detail how node-problem-detector get node name in README

### PR DESCRIPTION
**Pull request by andyxning**
Ref: #68

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/node-problem-detector/72)

### CODE REVIEW

**Comment by andyxning** [comment]
@Random-Liu PTAL.

File: README.md [hunk 1]
```diff
@@ -62,6 +62,7 @@ For example, to run without auth, use the following config:
 http://APISERVER_IP:APISERVER_PORT?inClusterConfig=false
 ```
 Refer [heapster docs](https://github.com/kubernetes/heapster/blob/1e40b0f4b5eeb3f02e11ee22c2b6fda36b6e6ea1/docs/source-configuration.md#kubernetes) for a complete list of available options.
+* `-hostname-override`: A customized node name used for node-problem-detector to update conditions and emit events. node-problem-detector gets node name first from `hostname-override`, then `NODE_NAME` environment variable and finally fall back to `os.Hostname` when node name can not be inferred from neither `hostname-override` nor `NODE_NAME` environment variable.
```
Comment (Random-Liu) [inline README.md]: `when node name can not be inferred from neither `hostname-override` nor `NODE_NAME` environment variable` seems unnecessary to me.

File: cmd/node_problem_detector.go [hunk 2]
```diff
@@ -41,6 +42,36 @@ func validateCmdParams() {
 	}
 }
 
+func mustGetNodeName() string {
+	var nodeName string
+
+	// Check hostname override first for customized node name.
+	if *hostnameOverride != "" {
```
Comment (Random-Liu) [inline cmd/node_problem_detector.go]:
```
if *hostnameOverride != "" {
  return *hostnameOverride
}
```

File: cmd/node_problem_detector.go [hunk 3]
```diff
@@ -41,6 +42,36 @@ func validateCmdParams() {
...
+	}
+
+	// Get node name from environment variable NODE_NAME
+	// By default, assume that the NODE_NAME env should have been set with
+	// downward api or user defined exported environment variable. We prefer it because sometimes
+	// the hostname returned by os.Hostname is not right because:
+	// 1. User may override the hostname.
+	// 2. For some cloud providers, os.Hostname is different from the real hostname.
+	nodeName = os.Getenv("NODE_NAME")
+	if nodeName == "" {
```
Comment (Random-Liu) [inline cmd/node_problem_detector.go]: Less indent is preferred. My original code is not good. :)
```
nodeName := os.Getenv("NODE_NAME")
if nodeName != "" {
  return nodeName
}

nodeName, err := os.Hostname()
if err != nil {
  panic("Failed to get host name: %v", err)
}

return nodeName
```

File: cmd/node_problem_detector.go [hunk 4]
```diff
@@ -41,6 +42,36 @@ func validateCmdParams() {
 	}
 }
 
+func mustGetNodeName() string {
```
Comment (Random-Liu) [inline cmd/node_problem_detector.go]: We usually use `OrDie` for function which may panic,`getNodeNameOrDie`

File: cmd/node_problem_detector.go [hunk 5]
```diff
@@ -50,7 +81,9 @@ func main() {
 		os.Exit(0)
 	}
 
+	nodeName := mustGetNodeName()
+
 	k := kernelmonitor.NewKernelMonitorOrDie(*kernelMonitorConfigPath)
-	p := problemdetector.NewProblemDetector(k, *apiServerOverride)
+	p := problemdetector.NewProblemDetector(k, *apiServerOverride, nodeName)
```
Comment (Random-Liu) [inline cmd/node_problem_detector.go]: Can we create the client here, and pass the client into `problemdetector` so that we don't need to pass these variables all the way down?

File: README.md (hunk 1 above)
Comment (andyxning) [inline README.md]: It is indeed verbose. :)

File: cmd/node_problem_detector.go (hunk 2 above)
Comment (andyxning) [inline cmd/node_problem_detector.go]: Done.

File: cmd/node_problem_detector.go (hunk 4 above)
Comment (andyxning) [inline cmd/node_problem_detector.go]: Having thought that `must` prefix is a convention for function which may panic in Golang. Since `OrDie` is the convention for kubernetes, we should stick with it. :)

File: cmd/node_problem_detector.go (hunk 3 above)
Comment (andyxning) [inline cmd/node_problem_detector.go]: Refactored

Comment (andyxning) [inline cmd/node_problem_detector.go]: Refactored

File: cmd/node_problem_detector.go (hunk 5 above)
Comment (andyxning) [inline cmd/node_problem_detector.go]: @Random-Liu I agree with you.

BTW, as we may add more command line argument, we may need to add a `Conf` struct and make all command line arguments as its fields. When we need a command line argument, we just need to reference an `Conf` instance. This will also make the code more readable. :)

I suggest we make another PR to make this refactor. WDYT.

Comment (andyxning) [inline cmd/node_problem_detector.go]: @Random-Liu I agree with you.

BTW, as we may add more command line argument, we may need to add a `Conf` struct and make all command line arguments as its fields. When we need a command line argument, we just need to reference an `Conf` instance. This will also make the code more readable. :)

I suggest we make another PR to make this refactor. WDYT.

Comment (andyxning) [inline cmd/node_problem_detector.go]: @Random-Liu I agree with you.

BTW, as we may add more command line argument, we may need to add a `Conf` struct and make all command line arguments as its fields. When we need a command line argument, we just need to reference an `Conf` instance. This will also make the code more readable. :)

I suggest we make another PR to make this refactor. WDYT.

File: cmd/node_problem_detector.go (hunk 4 above)
Comment (Random-Liu) [inline cmd/node_problem_detector.go]: @andyxning Thanks.

File: cmd/node_problem_detector.go (hunk 5 above)
Comment (Random-Liu) [inline cmd/node_problem_detector.go]: Yeah, I agree.

In fact, following the K8s convention, we should add an `option/` directory under `cmd/` and define the config object and flags there.

Feel free to file an issue or send a PR. :)

**Comment by Random-Liu** [comment]
@andyxning LGTM. Thanks!

**Comment by Random-Liu** [comment]
/lgtm

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/node-problem-detector/72#-:-KamxP2fXsTPqIKLJ3Kf:b-8ks5qa)*

File: cmd/node_problem_detector.go (hunk 5 above)
Comment (andyxning) [inline cmd/node_problem_detector.go]: @Random-Liu Will do this later.
